### PR TITLE
feat(library): add Polygraf PII detection and masking integration

### PR DESCRIPTION
**Pull request by ali-khoda**
- Fix Polygraf API auth/payload: Authorization: Bearer $POLYGRAF_API_KEY + request flags; robust response parsing.

- Align Polygraf docs/examples with the working cloud endpoint + entity labels (Email, Person, Phone).

- Local validation: pytest -q tests/test_polygraf.py passed.

## Release Notes

* **New Features**
  * Added Polygraf integration for detecting and masking personally identifiable information (PII) across user input, bot output, and retrieval sources.
  * Supports configurable entity type detection with options for cloud and self-hosted deployments.

* **Documentation**
  * Added comprehensive Polygraf integration guide with configuration examples for PII detection and masking workflows.

[Review Change Stack](https://app.coderabbit.ai/change-stack/NVIDIA-NeMo/Guardrails/pull/1693)

### CODE REVIEW

**Comment by DataMonarch** [comment]
Updated the PR to address the review feedback and keep the diff focused on the Polygraf integration.

Changes made:
- Added `**kwargs` support to `polygraf_mask_pii`.
- Added a warning when `POLYGRAF_API_KEY` is not set, so cloud auth failures are easier to diagnose.
- Added optional shared `aiohttp.ClientSession` support to `polygraf_request` and passed it through from actions.
- Expanded Polygraf tests for request payload/auth behavior, wrapped responses, error handling, missing API-key diagnostics, and extra kwargs/session passthrough.
- Restored unrelated core files to match current `develop`, so the PR no longer includes embedding/KV/model-engine/dependency changes.

Validation:
- `ruff check` passed for touched Polygraf files.
- `ruff format --check` passed for touched Polygraf files.
- `pytest tests/test_polygraf.py -v`: 13 passed.
- `pytest tests/test_gliner.py tests/test_rails_llm_config.py tests/test_config_loading.py -q`: 37 passed.
- Additional smoke checks passed for detect/mask behavior, request payload/header behavior, shared session passthrough, and v1/v2 Polygraf import loading.

**Comment by DataMonarch** [comment]
Addressed the latest Greptile/CodeRabbit follow-up comments.

Updates:
- Treat wrapped Polygraf responses with `entities: null` as an empty result.
- Preserve JSON parsing exception context with exception chaining.
- Tightened masking assertions and fixed the malformed completion string in tests.
- Added a concise retrieval-flow example to the Polygraf docs and clarified entity label guidance.

Validation:
- `ruff check` passed for touched code files.
- `ruff format --check` passed for touched code files.
- `pytest tests/test_polygraf.py tests/test_gliner.py tests/test_rails_llm_config.py tests/test_config_loading.py -q`: 51 passed.

**Comment by DataMonarch** [comment]
Pushed self-review hardening for two production risks I found.

- Added a default 30s aiohttp `ClientTimeout` for Polygraf requests. Without this, a slow or unresponsive Polygraf endpoint could hang the rails pipeline indefinitely (sibling community integrations like policyai/jailbreak/ai_defense/crowdstrike all set timeouts).
- Made the detect/mask actions defensive against malformed Polygraf responses. Non-dict entries, entries with missing `entity_type`, or with non-integer offsets are now skipped with a warning instead of raising `KeyError`/`AttributeError` mid-rail.
- Added tests covering timeout propagation and malformed-entity skipping.

Validation:
- `ruff check` passed.
- `ruff format --check` passed.
- `pytest tests/test_polygraf.py tests/test_gliner.py tests/test_rails_llm_config.py tests/test_config_loading.py -q` → 53 passed.

Note: the earlier `pr-tests-matrix (Ubuntu, 3.11)` failure was an unrelated flaky perf threshold in `tests/v2_x/test_state_serialization.py` (`avg_time < 0.2` failed at 0.211s). It passes locally and on the other matrix entries.

File: tests/test_polygraf.py [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+              "I can't answer that."
+        """,
+    )
+
+    chat = TestChat(
+        config,
+        llm_completions=[
+            "  express greeting",
+            '  "Hi! I am John.',
+        ],
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply tests/test_polygraf.py]: is this resolved?

File: tests/test_polygraf.py [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+    )
+    chat.app.register_action(
+        create_mock_polygraf_mask_pii(["Email", "Person"]),
+        "polygraf_mask_pii",
+    )
+
+    chat >> "Hi!"
+    response = chat.app.generate(messages=[{"role": "user", "content": "Hi!"}])
+    assert "John" not in response["content"] or "<NAME>" in response["content"]
+
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply tests/test_polygraf.py]: is this resolved?

File: nemoguardrails/library/polygraf/request.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    async with session.post(server_endpoint, json=payload, headers=headers) as resp:
+        if resp.status != 200:
+            raise ValueError(f"Polygraf call failed with status code {resp.status}.\nDetails: {await resp.text()}")
+
+        try:
+            data = await resp.json()
+        except aiohttp.ContentTypeError:
+            raise ValueError(
+                f"Failed to parse Polygraf response as JSON. Status: {resp.status}, Content: {await resp.text()}"
+            )
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/request.py]: is this resolved?

File: docs/configure-rails/guardrail-catalog/community/polygraf.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
+# Polygraf
+
+[Polygraf](https://polygraf.ai/) offers a state-of-the-art PII Detection & Masking API designed to help identify and protect sensitive information in your text.
+
+This integration enables the NeMo Guardrails library to detect or mask PII on user input, bot output, and/or retrieved documents.
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply docs/configure-rails/guardrail-catalog/community/polygraf.md]: is this resolved?

File: nemoguardrails/library/polygraf/request.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+        # Polygraf may return either a raw list of entities or a wrapper object.
+        if isinstance(data, list):
+            return data
+        if isinstance(data, dict):
+            entities = data.get("entities")
+            if isinstance(entities, list):
+                return entities
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/request.py]: @greptileai this is resolved in the latest commits. Do you confirm?

File: nemoguardrails/library/polygraf/actions.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        valid_sources = ["input", "output", "retrieval"]
+        raise ValueError(
+            f"Polygraf can only be defined in the following flows: {valid_sources}. "
+            f"The current flow, '{source}', is not allowed."
+        )
+
+    enabled_entities = source_config.entities if source_config.entities else None
+    api_key = os.environ.get("POLYGRAF_API_KEY")
+
+    entities = await polygraf_request(text, server_endpoint, api_key)
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: @greptileai this is resolved in the latest commits. Do you confirm?

File: nemoguardrails/library/polygraf/request.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        "pid_granularity": 3,
+        "aggregate_entities": True,
+    }
+    headers: Dict[str, str] = {"Content-Type": "application/json"}
+
+    if api_key:
+        headers["Authorization"] = f"Bearer {api_key}"
+
+    async with aiohttp.ClientSession() as session:
+        async with session.post(server_endpoint, json=payload, headers=headers) as resp:
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/request.py]: @greptileai this is resolved in the latest commits. Do you confirm?

File: nemoguardrails/library/polygraf/actions.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+        return False
+
+    if enabled_entities:
+        return any(e["entity_type"] in enabled_entities for e in entities)
+
+    return True
+
+
+@action(is_system_action=False)
+async def polygraf_mask_pii(source: str, text: str, config: RailsConfig) -> str:
```
[the remark this answers is not in the document]
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: @greptileai this is resolved in the latest commits. Do you confirm?

**Comment by Pouyanpi** [comment]
@DataMonarch would you please rebase to latest develop branch and resolve the conflicts. It seems you've addressed all the review comments from coderabbit and greptile, once the conflicts are resolved I'll start the review. We will definitely have it for 0.23.0 release, sorry for the delays.

File: nemoguardrails/library/polygraf/actions.py [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    for entity in entities:
+        if not isinstance(entity, dict):
+            log.warning("Skipping non-dict Polygraf entity: %r", entity)
+            continue
+        entity_type = entity.get("entity_type")
+        start = entity.get("start")
+        end = entity.get("end")
+        if entity_type is None or not isinstance(start, int) or not isinstance(end, int):
+            log.warning("Skipping Polygraf entity with missing or invalid fields: %r", entity)
+            continue
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/actions.py]: we should fail close when a selected PII span is malformed?

when Polygraf reports selected PII but omits an offset or returns one with the wrong type, this branch silently skips that entity and returns the remaining text successfully. For mixed responses, valid spans are masked while the malformed span's raw PII is sent onward; fail the action or otherwise guarantee masking instead of continuing.

File: nemoguardrails/rails/llm/config.py [hunk 10]
```diff
@@ -392,6 +392,36 @@ class GLiNERDetection(BaseModel):
...
+        description="Configuration of the entities to be detected on the user input.",
+    )
+    output: PolygrafDetectionOptions = Field(
+        default_factory=PolygrafDetectionOptions,
+        description="Configuration of the entities to be detected on the bot output.",
+    )
+    retrieval: PolygrafDetectionOptions = Field(
+        default_factory=PolygrafDetectionOptions,
+        description="Configuration of the entities to be detected on retrieved relevant chunks.",
+    )
```
Comment (Pouyanpi) [inline nemoguardrails/rails/llm/config.py:426]: we should reject unknown Polygraf configuration keys.

add ConfigDict(extra="forbid") to both models.

File: nemoguardrails/library/polygraf/request.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+    timeout: aiohttp.ClientTimeout,
+) -> List[Dict[str, Any]]:
+    try:
+        post_ctx = session.post(server_endpoint, json=payload, headers=headers, timeout=timeout)
+    except TypeError:
+        # Some test doubles do not accept a `timeout` kwarg; fall back to the
+        # session-level timeout instead.
+        post_ctx = session.post(server_endpoint, json=payload, headers=headers)
+
+    async with post_ctx as resp:
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/request.py]: When the endpoint exceeds the configured timeout or DNS/connection setup fails, entering this context raises `TimeoutError` or an `aiohttp.ClientError`, bypassing the json only exception handler. This contradicts `polygraf_request` and the documentation's `ValueError` contract, so network failures should be caught around the request context and re-raised with the documented exception type.

File: nemoguardrails/library/polygraf/actions.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+    safe_entities = []
+    for entity in entities:
+        if not isinstance(entity, dict):
+            log.warning("Skipping non-dict Polygraf entity: %r", entity)
+            continue
+        entity_type = entity.get("entity_type")
+        start = entity.get("start")
+        end = entity.get("end")
+        if entity_type is None or not isinstance(start, int) or not isinstance(end, int):
+            log.warning("Skipping Polygraf entity with missing or invalid fields: %r", entity)
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/actions.py]: when a malformed entity still contains `entity_text`, such as an email with a missing offset, logging the entire dictionary with %r writes the PII this integration is meant to protect into application logs. Log only structural metadata or the invalid field names.

Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: and adapt the tests accordingly

File: nemoguardrails/library/polygraf/flows.co [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+flow polygraf detect pii on retrieval
+  """Check if the relevant chunks from the knowledge base have any PII using Polygraf."""
+  $has_pii = await PolygrafDetectPiiAction(source="retrieval", text=$relevant_chunks)
+
+  if $has_pii
+    bot inform answer unknown
+    abort
+
+
+#######################################################
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/flows.co]: please remove this marker

File: nemoguardrails/library/polygraf/flows.co
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/flows.co:1]: we should fix Colang 2 variable scoping. these flows read $user_message, $bot_message, and $relevant_chunks without declaring them global. Colang 2 consequently sends text: null, allowing PII or returning it unmasked. 

Please declare the globals before access for every detection and masking flow.

File: nemoguardrails/library/polygraf/flows.co [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+#### POLYGRAF PII MASKING RAILS ####
+
+# INPUT RAILS
+
+flow polygraf mask pii on input
+  """Mask any detected PII in the user input using Polygraf."""
+  $masked_input = await PolygrafMaskPiiAction(source="input", text=$user_message)
+
+  global $user_message
+  $user_message = $masked_input
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/flows.co]: move it before

` $masked_input = await PolygrafMaskPiiAction(source="input", text=$user_message)`

File: nemoguardrails/library/polygraf/flows.v1.co [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+define subflow polygraf detect pii on retrieval
+  """Check if the relevant chunks from the knowledge base have any PII using Polygraf."""
+  $has_pii = execute polygraf_detect_pii(source="retrieval", text=$relevant_chunks)
+
+  if $has_pii
+    bot inform answer unknown
+    stop
+
+
+#######################################################
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/flows.v1.co]: remove this

File: nemoguardrails/library/polygraf/actions.py [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+        raise ValueError(
+            f"Polygraf can only be defined in the following flows: {valid_sources}. "
+            f"The current flow, '{source}', is not allowed."
+        )
+
+    enabled_entities = source_config.entities if source_config.entities else None
+    api_key = _get_polygraf_api_key()
+    session = kwargs.get("session")
+
+    entities = await polygraf_request(text, server_endpoint, api_key, session=session)
```
Comment (Pouyanpi) [inline nemoguardrails/library/polygraf/actions.py]: Provider errors propagate into `ActionDispatcher`, whose exception path logs action parameters including text. A timeout can therefore expose raw PII. Please catch normalized request failures in these actions: block for detection and fully redact for masking, logging only a structural warning.

File: tests/test_polygraf.py
Comment (Pouyanpi) [inline tests/test_polygraf.py:1]: add test for colang 2.0 coverage

**Review by Pouyanpi — CHANGES_REQUESTED** [review]
@DataMonarch @ali-khoda 

Thanks for the PR, please have a look at following review comments. We are good to merge once these are addressed. Thank you.

**Comment by DataMonarch** [comment]
@Pouyanpi addressed your review.

Changes in `bbceb7e2e`:

**Fail-closed semantics (actions.py)**
- Provider/network failures from `polygraf_request` (now caught as `ValueError`) cause `polygraf_detect_pii` to return `True` and `polygraf_mask_pii` to return a new `FAILSAFE_MASK_PLACEHOLDER` (`<REDACTED>`). The original text never propagates downstream.
- If Polygraf reports a *configured* (selected) entity with missing/invalid offsets or type, both actions also fail closed. Unselected malformed entities are still skipped (no false-positive blocks).
- All warning logs were rewritten to use only structural metadata (`shape`, `keys`, `invalid_fields`); no entity text or response body is ever logged.

**Request layer (request.py)**
- `asyncio.TimeoutError` and `aiohttp.ClientError` are normalized to `ValueError`, so timeouts/DNS/TLS failures honor the documented exception contract.

**Config (config.py)**
- Both `PolygrafDetection` and `PolygrafDetectionOptions` now use `ConfigDict(extra="forbid")`.

**Colang 2 flows (flows.co)**
- Every flow now declares `global $user_message` / `$bot_message` / `$relevant_chunks` before reading the rail variable. Mask flows also place the `global` declaration above the action call so the assignment binds to the rail variable.
- Removed the stray `#######...` separator lines from `flows.co` and `flows.v1.co`.

**Tests (test_polygraf.py)**
- New coverage:
  - `test_polygraf_request_normalizes_timeout_as_value_error`
  - `test_polygraf_request_normalizes_client_error_as_value_error`
  - `test_polygraf_config_rejects_unknown_keys`
  - `test_polygraf_mask_pii_fails_closed_on_malformed_selected_entity` (asserts the PII value never appears in result or logs)
  - `test_polygraf_mask_pii_skips_unselected_malformed_entity`
  - `test_polygraf_mask_pii_fails_closed_on_provider_error`
  - `test_polygraf_detect_pii_fails_closed_on_provider_error`
  - `test_polygraf_detect_pii_fails_closed_on_malformed_selected_entity`
- Colang 2 coverage via parser:
  - `test_polygraf_v2_flows_parse_successfully`
  - `test_polygraf_v2_flows_declare_required_globals` (catches missing `global` regressions in `flows.co`)

**Note on the `Check Fern docs` failure**
The failure log shows `Authentication required. Please run 'fern login' or set the FERN_TOKEN environment variable.` This is `FERN_TOKEN` not being available to fork PRs (a standard GitHub Actions limitation), not a problem in this PR. Happy to defer to maintainer guidance.

**Comment by Pouyanpi** [comment]
Thank you @DataMonarch your main fix/changes work, would you please address following issues, these are not fully addressed yet.


1. malformed spans can still fail open at nemoguardrails/library/polygraf/actions.py.

    _classify_entities still treats some malformed spans as safe or unselected. When entity_type is missing under an enabled filter, has_malformed_selected remains false; reversed or
    out-of-range integer offsets are appended to safe_spans. These cases leave raw PII unchanged. Please fail closed when the type is unknown, reject booleans, and validate 0 <= start < end <= len(text).

2. documentation contract is stale at docs/configure-rails/guardrail-catalog/community/polygraf.mdx

    It says the integration raises ValueError, but actions now catch it and either block detection or return <REDACTED>.

3. Coverage gap at tests/test_polygraf.py.

    The new Colang 2 tests inspect parsed declarations but do not execute the rails. Actual local runtime verification passed, but a TestChat regression test would protect the behavior that originally produced text: null.

**Comment by DataMonarch** [comment]
@Pouyanpi thanks for catching the remaining gaps. Pushed `e0a5faf4d`:

**1. Malformed-span fail-open in `_classify_entities`** → fixed
- Missing/empty `entity_type` now fails closed (we cannot tell whether the entity would have matched the filter).
- `bool` offsets explicitly rejected (`isinstance(True, int)` is True in Python, so a plain int check would pass them through).
- Added `0 <= start < end <= len(text)` validation; reversed/negative/empty/out-of-range spans no longer reach `safe_spans`. If such a span matched the filter (or the filter is `None`), the action fails closed.
- `_classify_entities` now takes `text_length` so it can do the range check.

**2. Stale `ValueError` contract in `polygraf.mdx`** → fixed
- Removed the "raises `ValueError`" note. Added a `## Failure Handling` section that documents the actual contract: detect blocks (`True`), mask returns `<REDACTED>`, and what triggers each fail-closed path (provider/network/parse failure, malformed entity span, missing/invalid offsets). Structural-only logging is also called out.

**3. Coverage gap (parser-only Colang 2 tests)** → fixed
- New `test_polygraf_v2_input_flow_passes_actual_user_message_to_action` is a `TestChat`-based regression. It loads the shipped `flows.co` text, wires it into a v2 input rail via `flow input rails $input_text`, registers a mock `polygraf_mask_pii` that records the `text` it received, then sends a user message. It asserts the action received the actual user text rather than `None`. Verified locally that removing the `global $user_message` declaration makes the test fail with the exact bug you described.
- Additional malformed-span tests: missing entity_type (mask + detect), and a parametrized test over bool start, bool end, negative start, reversed offsets, end past text length, and empty spans.

Local validation: ruff clean, all 34 Polygraf tests pass.

File: nemoguardrails/library/polygraf/actions.py (hunk 9 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: @coderabbit do you confirm this is resolved?

File: nemoguardrails/rails/llm/config.py (hunk 10 above)
Comment (Pouyanpi) [reply nemoguardrails/rails/llm/config.py:426]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/request.py (hunk 11 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/request.py]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/actions.py (hunk 12 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/flows.co (hunk 13 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/flows.co]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/flows.co
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/flows.co:1]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/flows.co (hunk 14 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/flows.co]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/flows.v1.co (hunk 15 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/flows.v1.co]: @CodeRabbit do you confirm this is resolved?

File: nemoguardrails/library/polygraf/actions.py (hunk 16 above)
Comment (Pouyanpi) [reply nemoguardrails/library/polygraf/actions.py]: @CodeRabbit do you confirm this is resolved?

File: tests/test_polygraf.py
Comment (Pouyanpi) [reply tests/test_polygraf.py:1]: @CodeRabbit do you confirm this is resolved?

**Review by Pouyanpi — APPROVED** [review]
Thanks @DataMonarch for making the changes and your flexibility 👍🏻 merging 🚀
